### DISCUSSION TITLE
Feedback to upgrade jiff to 0.1.16

### DISCUSSION THREAD

**Discussion by tisonkun** [BurntSushi/jiff#174] [Experience Reports]
Hi @BurntSushi! Thanks for your continuous work and the recent https://github.com/BurntSushi/jiff/commit/f6fc1b07e9b6fb0fdae675b158cc5b5f59d205cb and https://github.com/BurntSushi/jiff/commit/1e39b60d7b811f9eba3ab77d4aacd890cbb5a34a released in 0.1.16.

I have upgraded jiff to 0.1.16 for our database software. Here is the patch to demonstrate that the changeset (from 0.1.13) is relatively small: https://gist.github.com/tisonkun/183e761fdf8d26a4799f91f6ba3cbe80

Briefly, we use SignedDuration as the internal representation of the database's [`Interval`](https://www.scopedb.io/reference/datatypes-datetime) type, and now try to use it in the HTTP request to reduce the dependency to humantime/humantime-serde.

Here are three remaining issues I'm considering now.

1. Converting SignedDuration to std's Duration introduces some boilerplate code.

Although I think this is by design (std's Duration is unsigned), the `.try_into.map_err`/`.try_into.except` is a bit noisy.

Ecosystems like `tokio` don't integrate with `jiff`'s type (yet?), so we have to manage such boilerplate code in the application.

2. Support `Span` as the representation of a new interval type.

We use `SignedDuration` as the internal representation because it has a natural mapping from/to nanoseconds. While with `Span`, we may need to store the whole string or a few of accurate parts. In the upside, `Span` allows user to specify intervals with days, months, years, etc., which can be more intuitive in some case.

Software like Apache Arrow has multiple types for duration and interval and that may be the way to go.

* https://docs.rs/arrow/latest/arrow/datatypes/struct.IntervalMonthDayNano.html
* https://docs.rs/arrow/latest/arrow/array/type.DurationNanosecondArray.html

3. Eliminate `sqlx_jiff` internal wrapper.

This is unrelated to 0.1.16 but we're now internally maintain a mod to do the same thing exactly for what https://github.com/BurntSushi/jiff/pull/141 does. If it can get some reviews, it would be great :P

Thanks again for this wonderful crate. I find it quite ergonomic and has the minimum friction to use.

**Comment by tisonkun** [comment]
For anyone who would like to upgrade to 0.1.16 and have an option to switch back the previous format, check out https://docs.rs/jiff/latest/jiff/fmt/temporal/struct.SpanPrinter.html.

**Reply by BurntSushi** [reply]
> For anyone who would like to upgrade to 0.1.16 and have an option to switch back the previous format, check out https://docs.rs/jiff/latest/jiff/fmt/temporal/struct.SpanPrinter.html.

You shouldn't need to use this unless you specifically only want to support the ISO 8601 format. Otherwise there is no change to what Jiff does in semver APIs other than FromStr and Deserialize accepting more inputs.

**Reply by tisonkun** [reply]
I agree with your statement. My assumption is, now the format of `SignedDuration`, said in `{dur:?}` becomes `59s 778ms 800µs` and `26h 59s 778ms 800µs ago`. When people may want to print `PT59.7788s` and `-PT26h59.7788s`, `temporal::SpanPrint` is an exit. Or perhaps `{dur}` does the same?

**Reply by BurntSushi** [reply]
`{dur:?}` is the Debug format. Like with std, there aren't any API guarantees about its specific output format. In contrast, the Display impl format is an API guarantee. So `{dur}` is always guaranteed to be ISO 8601 (in a semver compatible release, although I don't expect to change it in future semver incompatible releases).

**Reply by tisonkun** [reply]
So if I'd like to use the friendly format stably, instead of depending on `{dur:?}`, should I use `friendly::SpanPrinter`? Or if `{dur:#}` is promised not to change in a compatible version bump.

**Reply by BurntSushi** [reply]
`{dur:#}` being the friendly format is a semver guarantee. Display impls are supposed to be stable parts of APIs.

You really should only need `friendly::SpanPrinter` when you want to customize the output or have more control over what it writes to.

**Comment by BurntSushi** [comment]
For (3), I consider that part of ecosystem integration. My focus hasn't really shifted to that yet.

For (2), is there any actionable feedback there? I'm not clear on what you're saying exactly. :)

For (1), do you have any alternative suggestions? The conversion from SignedDuration to Duration is fallible. So there has to be an error path in there somewhere.

**Reply by tisonkun** [reply]
> For (3), I consider that part of ecosystem integration. My focus hasn't really shifted to that yet.

Understandable. I'm fine with maintaining the one file glue in my codebase.

> For (2), is there any actionable feedback there? I'm not clear on what you're saying exactly. :)

Just sharing how a database, or generally data processing systems, would support and handle interval (span) and duration type. Nothing needs to be done in jiff side.

> For (1), do you have any alternative suggestions? The conversion from SignedDuration to Duration is fallible. So there has to be an error path in there somewhere.

Nope. I think this is by design and adding several checks or unwrap known value is fine. Actually, if I use std's `Duration` + humantime_serde, the check is done during serde.

**Reply by tisonkun** [reply]
> For (2), is there any actionable feedback there? I'm not clear on what you're saying exactly. :)

Oh yes. There is one thing I need help: if I'd store some data to reconstruct a `Span`, what is the minimum? Said Apache Arrow has a type `IntervalMonthDayNano`, is it possible I convert a `Span` to `(months, days, nanos)` and later reconstruct the `Span` accurately?

That said, storing the string representation of `Span` can be expensive than a few integers.

**Reply by BurntSushi** [reply]
No. The only way to losslessly serialize a `Span` is to preserve every individual unit. That was one of the core motivations for the friendly format (which is explained in the friendly module docs). Indeed, that is the _entire point_ of it. If you don't need to preserve individual units, then you are more than welcome to use a lossy representation (which even ISO 8601 is). It all depends on your requirements. But if you don't capture each individual unit, some distinct `Span` values will invariable serialize to the same data, thus making it ambiguous. That _could_ be fine. It depends on your requirements.

**Reply by BurntSushi** [reply]
You can invent a binary serialization format, to be clear, but it should capture every unit value.

**Reply by BurntSushi** [reply]
After thinking about this more, I think a clarification is important: you _can_ use `(months, days, nanos)` to represent a `Span` if you don't care about the difference between `36 months` and `3 years`. They are precisely equivalent durations in terms of real time passed, but they might express different flavors of time passage. For example, you might want to always represent the length of films in terms of minutes, even if exceeds multiple hours, as is convention in the US at least. The only way to preserve that expressive power is to serialize all units as they are.

It all depends on your requirements.

**Reply by tisonkun** [reply]
Thanks for your information.

> They are precisely equivalent durations in terms of real time passed.

Yes this is the point I'd like to confirm.

> you might want to always represent the length of films in terms of minutes, even if exceeds multiple hours, as is convention in the US at least

Good to know.

The rest part would be how such `Span`s get compared and checked equality. I found there are several docs to read the test, but generally, (1) equality is checked by exact field matches (as derived), and (2) compare is done faillable, depending on the unit in use. Is this the case?

**Reply by BurntSushi** [reply]
The compare method uses the actual time passed. 

The Eq and PartialEq impls will be removed in jiff 0.2 since they are a footgun. There's an open issue about it.

**Reply by tisonkun** [reply]
Get it.

For cross reference - https://github.com/BurntSushi/jiff/issues/32.
